tests/framework/extensions/namespaces/namespaces.go

```diff
@@ -0,0 +1,100 @@
+package namespaces
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/rancher/rancher/pkg/api/scheme"
+	"github.com/rancher/rancher/tests/framework/clients/rancher"
+	"github.com/rancher/rancher/tests/framework/pkg/convert"
+	"github.com/rancher/rancher/tests/framework/pkg/wait"
+	"github.com/rancher/rancher/tests/integration/pkg/defaults"
+	coreV1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/watch"
+)
+
+var NamespaceGroupVersionResource = schema.GroupVersionResource{
+	Group:    "",
+	Version:  "v1",
+	Resource: "namespaces",
+}
+
+// ContainerDefaultResourceLimit sets the container default resource limit in a string
+// limitsCPU and requestsCPU in form of "3m"
+// limitsMemory and requestsMemory in the form of "3Mi"
+func ContainerDefaultResourceLimit(limitsCPU, limitsMemory, requestsCPU, requestsMemory string) string {
+	containerDefaultResourceLimit := fmt.Sprintf("{\"limitsCpu\": \"%s\", \"limitsMemory\":\"%s\",\"requestsCpu\":\"%s\",\"requestsMemory\":\"%s\"}",
+		limitsCPU, limitsMemory, requestsCPU, requestsMemory)
+	return containerDefaultResourceLimit
+}
+
+// NewProjectNamespace is a constructor that sets up a namespace object to be used by the GetDownStreamK8Client to create a project namespace
+func NewProjectNamespace(namespaceName, projectID, containerDefaultResourceLimit string, labels, annotations map[string]string) *coreV1.Namespace {
+	annotations["field.cattle.io/containerDefaultResourceLimit"] = containerDefaultResourceLimit
```

Comment:
im confused is this functon creating a namespace in a project?

Reply:
No this just the constructor for the project namespace. The function below actually creates it, and registers a cleanup func for it.